### ISSUE TITLE
nodeinit.restartPods=true isn't working on GKE 

### ISSUE THREAD

**Issue by racarvalho1987** [cilium/cilium#15177]
{
textPayload: "find: '/var/lib/cni/networks/': No such file or directory
"
resource: {
type: "k8s_container"
labels: {
project_id: "xxxxx"
location: "xxxxxx"
pod_name: "cilium-node-init-rqzcf"
cluster_name: "gke-cluster-xxxxx"
namespace_name: "cilium"
container_name: "node-init"
}
}
timestamp: "2021-03-02T12:25:46.033791028Z"
severity: "ERROR"
labels: {
k8s-pod/app: "cilium-node-init"
k8s-pod/pod-template-generation: "1"
k8s-pod/controller-revision-hash: "xxxxx"
compute.googleapis.com/resource_name: "gke-xxxxxx"
}
logName: "projects/xxxxxxx/logs/stderr"
receiveTimestamp: "2021-03-02T12:25:55.026224571Z"
}

- Cilium version: 1.9.0
- GKE version: 1.17.15-gke.800

1. Create a GKE cluster
2. Install Cilium with nodeinit.restartPods=true

`helm install --wait --timeout 300s cilium cilium/cilium --version 1.9.0 --namespace cilium --set nodeinit.enabled=true --set nodeinit.reconfigureKubelet=true --set nodeinit.removeCbrBridge=true --set nodeinit.restartPods=true --set cni.binPath=/home/kubernetes/bin --set gke.enabled=true --set ipam.mode=kubernetes --set nativeRoutingCIDR=$(gcloud container clusters describe gke-cluster-xxx --zone xxxxxx --format 'value(clusterIpv4Cidr)') --set kubeProxyReplacement=disabled --set hubble.listenAddress=':4244'  --set hubble.relay.enabled=true --set hubble.ui.enabled=true `

3. gcloud container clusters resize gke-cluster-xxxx --node-pool gke-cluster-cbugscaleup-node-pool --num-nodes 0
    wait until all pods terminate
4. gcloud container clusters resize gke-cluster-xxxx --node-pool gke-cluster-cbugscaleup-node-pool --num-nodes 1
    "find: '/var/lib/cni/networks/': No such file or directory during node startup

So, everytime that I need to do resize nodes I need to manually restart unmanaged pods.

**Comment by christarazi** [comment]
Thanks for filing the issue. I took a look at this and I was able to reproduce it with the latest version of Cilium as well (v1.9.5).

Strangely enough though, that directory does exist. If I do an `ls /var/lib/cni/networks/` before the `find`, it works. It seems like GKE is lazily mounting this path on first access. I don't know if that's what's happening for a fact, but it's behaving as such.

**Comment by christarazi** [comment]
Could you try applying this diff and see if that resolves it for you?

```
diff --git a/install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml b/install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
index 50cd92ba2..db98ad10c 100644
--- a/install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
+++ b/install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
@@ -225,6 +225,12 @@ spec:
               if [ ! -f /etc/crictl.yaml ] || grep -q 'docker' /etc/crictl.yaml; then
                 # Works for COS, ubuntu
                 # Note the first line is the containerID with a trailing \r
+                set +e
+                until [[ -d /var/lib/cni/networks ]]; do
+                  echo "waiting for /var/lib/cni/networks..."
+                  sleep 3s
+                done
+                set -e
                 for f in `find /var/lib/cni/networks/ -type f ! -name lock ! -name last_reserved_ip.0`; do docker rm -f "$(sed 's/\r//;1q' $f)" || true; done
               elif [ -n "$(docker ps --format '{{ "{{" }}.Image{{ "}}" }}' | grep ^[0-9]*\.dkr\.ecr\.[a-z]*-[a-z]*-[0-9]*\.amazonaws\.com/amazon-k8s-cni)" ]; then
                 timeout=1
```

**Comment by splushii** [comment]
Hitting this as well.

@christarazi using the loop waiting for `/var/lib/cni/networks` works! Unfortunately there is another race condition, as not all files/containers under `/var/lib/cni/networks/` may be present when the `... docker rm -f ...` runs. An infinite loop solves this (see patch below), but it also means that the script is stuck here (so the `restartPods` functionality must be at the end of the node-init script if this approach is used). Using this script, we eventually get no Cilium unmanaged nodes (even in the extreme scenario of scaling all nodes to 0 and then scaling up again).

Sometimes Cilium is faster than GKE (so that we **don't** have Cilium unmanaged Pods), and then the script ends up looping on `"waiting for /var/lib/cni/networks..."`. Other times GKE is faster than Cilium (so that we **do** have Cilium unmanaged Pods), and then the script ends up looping on `"Cleaning up Cilium unmanaged Pods..."` instead.

Seems like entries under `/var/lib/cni/networks` are propagated (e.g. we get Cilium unmanaged Pods) until the cilium-agent on the node is up and running. So it's also important that enough resources are reserved for the cilium-agents, otherwise it may take a long time as it competes with other workloads (being constantly restarted by the node-init script) for resources.

```patch
diff --git a/install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml b/install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
index 50cd92ba2..7a3d0e940 100644
--- a/install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
+++ b/install/kubernetes/cilium/templates/cilium-nodeinit-daemonset.yaml
@@ -224,8 +224,22 @@ spec:
               echo "Restarting kubenet managed pods"
               if [ ! -f /etc/crictl.yaml ] || grep -q 'docker' /etc/crictl.yaml; then
                 # Works for COS, ubuntu
-                # Note the first line is the containerID with a trailing \r
-                for f in `find /var/lib/cni/networks/ -type f ! -name lock ! -name last_reserved_ip.0`; do docker rm -f "$(sed 's/\r//;1q' $f)" || true; done
+                set +e
+                until [[ -d /var/lib/cni/networks ]]; do
+                  echo "waiting for /var/lib/cni/networks..."
+                  sleep 3s
+                done
+                set -e
+                while :; do
+                  echo "Cleaning up Cilium unmanaged Pods..."
+                  # Note the first line is the containerID with a trailing \r
+                  for f in `find /var/lib/cni/networks/ -type f ! -name lock ! -name last_reserved_ip.0`; do
+                    docker rm -f "$(sed 's/\r//;1q' $f)" || true
+                    rm -f "$f"
+                  done
+                  echo "Cleaning up Cilium unmanaged Pods done"
+                  sleep 10s
+                done
               elif [ -n "$(docker ps --format '{{ "{{" }}.Image{{ "}}" }}' | grep ^[0-9]*\.dkr\.ecr\.[a-z]*-[a-z]*-[0-9]*\.amazonaws\.com/amazon-k8s-cni)" ]; then
                 timeout=1
                 for i in $(seq 1 7); do
```

**Comment by christarazi** [comment]
@splushii Thanks for the detailed write-up. Would you mind opening a PR with that patch? We can figure out a solution to resolve the infinite loop problem during the review process.

**Comment by Champ-Goblem** [comment]
@christarazi @splushii 
I have noticed the same issue and have been testing a patch for this which seems to be pretty effective. I opted to add a wait in before the pods are restarted that checks if the cilium agent is ready by curling the socket it deploys. While cilium agent is unavailable the curl command times out and returns a non 0 exit code, once the socket is available the curl returns a zero exit code and allows the script to continue to restart the pods:
```
{{- if .Values.nodeinit.restartPods }}
{{- if .Values.nodeinit.waitForCiliumAgent }}
              echo "Waiting for cilium agent to be ready"
              set +e
              curl -X GET --unix-socket /run/cilium/cilium.sock http://localhost/healthz 1>/dev/null 2>/dev/null
              while [ $? -ne 0 ]; do
                sleep 1
                curl -X GET --unix-socket /run/cilium/cilium.sock http://localhost/healthz 1>/dev/null 2>/dev/null
              done
              set -e
              echo "Cilium agent ready"
{{- end }}
...restart logic...
{{- end }}
```
I thought it best to make this optional by the helm chart value `nodeinit.waitForCiliumAgent `.
There is one drawback, curling the current endpoint returns a 404 once the actual socket is available, maybe someone with some more knowledge about this could get it to return 200, then we can ensure that the health of the agent is good before continuing.

**Comment by christarazi** [comment]
@aanm is working on a PR to fix #16602 which mostly deprecate the use of nodeinit so this issue is likely to be resolved once the PR is up.
